pkg/syncutil/opadataclient.go

```diff
@@ -19,51 +19,20 @@ import (
 	"context"
 
 	"github.com/open-policy-agent/frameworks/constraint/pkg/types"
-	"github.com/open-policy-agent/gatekeeper/v3/pkg/watch"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
+// CacheManagerMediator is an interface for mediating
```

Comment:
the CMM is a mediator to solve the import cycle dependency. Should I add this in the comment?

Reply:
> To preserve coherence, all stuff related to syncsets should be moved to a different package.

pkg/controller/sync ?

> What is the import cycle?

`cachemanager` needs a reference to `syncc` to start the sync controller on integ tests and `syncc` needs a reference to `cachemanager` to manipulate the cache.

Reply:
> pkg/controller/sync ?

I think putting this code under `/controller/` was proposed before and rejected.

`pkg/cachemanager`
`pkg/cachemanager/aggregator`
`pkg/cachemanager/parser`

?

> cachemanager needs a reference to syncc to start the sync controller on integ tests and syncc needs a reference to cachemanager to manipulate the cache.

If we want to test the sync controller, the tests should go next to the sync controller. In practice, integration tests would probably be best placed next to the config controller and/or the SyncSet controller, since that's where the top of the chain is.

If we want to test cache manager specifically, we should be checking that functionality directly.

If using sync_controller makes it easier to write those tests (not recommended b/c of the import cycle), one way to break that would be to call the test package `cachemanager_test` (you will no longer have access to private methods in cache manager)

Reply:
The CacheManager does a lot of stuff, so some testing seems appropriate.  I can understand the desire to couple it with the sync_c when testing due to their collaborative role in the product.  That said, combining components into a running piece of software is accomplished well by e2e testing (which we should certainly leverage once we have both syncset_c and config_c).

A balanced distribution of testing approaches (unit, functional/integration/smoke, e2e) would be helpful here.

My advice to Alex is to find pieces of functionality that can be relatively easily set up and tested cheaply.  Then, rely on e2e for its strength in composition to verify that those elements are being combined successfully.  If I am very confident in my components, I can be less exhaustive in testing their composition (which has many permutations).  I care more that they _are_ successfully composed and that a basic user journey is functioning.

In the case of the CM, we can test it in isolation (no sync_c) in a unit test.  Finding ways to test less than 100% of this code could be valuable, as such tests could be low cost but provide significant (though not perfect) coverage.  

One strategy I often employ in this kind of situation is finding ways to separate logic that relies on tickers and channel blocking from logic that makes decisions (conditional logic) and manipulates stateful data structures.  Doing that could allow one to write a test that confirms that the correct decisions are made based on specific pre-conditions (these GVKs exist, this is the state of the cache, these GVKs are in GVKAggregator) and the correct resulting state is attained.

If it's hard to find things like that to test, that can sometimes be an indicator that the separation of concerns between parts of the code should be revisited.

Reply:
i've moved away from using the `mediator` interface and went with integration tests that use the sync_c controller in a `cachemanager_test`.

I think we have a pretty good mix of units and integs and now the distinctions is pretty clear but please let me know if there's things I can improve on regarding this code.